Add nest climate support for SDM API

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add client for nest SDM API.  This dispatches between API versions depending on how HA was configured.

This is an initial version, which still may need some polishing to be fully featured complete.  This uses a pubsub subscriber to update the entities, similar to the other types in the SDM API.  It supports cool, heat, heat_cool, and an eco mode -- depending on the traits supported by the device.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

allenporter committed:
Add Nest climate implementation for SDM API

allenporter committed:
Simplify update states.

allenporter committed:
Add fan support to nest climate

allenporter committed:
Remove dead return

allenporter committed:
Update nest climate comments to fix references

allenporter committed:
Use latest nest library

allenporter committed:
Fix nest pylint errors for no-else-return

mike9011 commented:
I'm getting an error testing this PR, maybe because we don't have a `FAN_MODE` in Europe?


```2020-11-02 10:38:42 ERROR (MainThread) [homeassistant.components.climate] Error adding entities for domain climate with platform nest
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 314, in async_add_entities
    await asyncio.gather(*tasks)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 505, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 531, in add_to_platform_finish
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 296, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 322, in _async_write_ha_state
    attr.update(self.state_attributes or {})
  File "/usr/src/homeassistant/homeassistant/components/climate/__init__.py", line 256, in state_attributes
    data[ATTR_FAN_MODE] = self.fan_mode
  File "/config/custom_components/nest/climate_sdm.py", line 265, in fan_mode
    return FAN_MODE_MAP.get(trait.timer_mode, FAN_OFF)
  File "/usr/local/lib/python3.8/site-packages/google_nest_sdm/device_traits.py", line 49, in timer_mode
    return self._data[TIMER_MODE]
KeyError: 'timerMode'
2020-11-02 10:38:42 ERROR (MainThread) [homeassistant.components.climate] Error while setting up nest platform for climate
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 205, in _async_setup_platform
    await asyncio.gather(*pending)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 314, in async_add_entities
    await asyncio.gather(*tasks)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 505, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 531, in add_to_platform_finish
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 296, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 322, in _async_write_ha_state
    attr.update(self.state_attributes or {})
  File "/usr/src/homeassistant/homeassistant/components/climate/__init__.py", line 256, in state_attributes
    data[ATTR_FAN_MODE] = self.fan_mode
  File "/config/custom_components/nest/climate_sdm.py", line 265, in fan_mode
    return FAN_MODE_MAP.get(trait.timer_mode, FAN_OFF)
  File "/usr/local/lib/python3.8/site-packages/google_nest_sdm/device_traits.py", line 49, in timer_mode
    return self._data[TIMER_MODE]
KeyError: 'timerMode'
```

allenporter commented:
Thanks, super helpful.

So the device has a [FanTrait](https://developers.google.com/nest/device-access/traits/device/fan#response), but not a timerMode in the output.  Any chance you could run the google_nest command line tool and dump the output? e.g.
```
$ google_nest --project_id="${PROJECT_ID}" list_devices
```
Or generally the output of the device if using another means like curl.  You can scrub any device ids, I really am just interested in the traits output.

I can put in a fix given what you've said so far, but want to make sure I'm not missing anything else

allistermaguire commented:
Hi Allen, appreciate all your work getting the Nest SDM working. I have European versions of the Nest Thermostat so can provide you with a device output.

```
{
  "name": "enterprises/XXXXXXXX/devices/XXXXXXXX",
  "type": "sdm.devices.types.THERMOSTAT",
  "assignee": "enterprises/XXXXXXXX",
  "traits": {
	"sdm.devices.traits.Info": {
	  "customName": ""
	},
	"sdm.devices.traits.Humidity": {
	  "ambientHumidityPercent": 55
	},
	"sdm.devices.traits.Connectivity": {
	  "status": "ONLINE"
	},
	"sdm.devices.traits.Fan": {},
	"sdm.devices.traits.ThermostatMode": {
	  "mode": "HEAT",
	  "availableModes": [
		"HEAT",
		"OFF"
	  ]
	},
	"sdm.devices.traits.ThermostatEco": {
	  "availableModes": [
		"OFF",
		"MANUAL_ECO"
	  ],
	  "mode": "OFF",
	  "heatCelsius": 15.08165,
	  "coolCelsius": 24.44443
	},
	"sdm.devices.traits.ThermostatHvac": {
	  "status": "HEATING"
	},
	"sdm.devices.traits.Settings": {
	  "temperatureScale": "CELSIUS"
	},
	"sdm.devices.traits.ThermostatTemperatureSetpoint": {
	  "heatCelsius": 20
	},
	"sdm.devices.traits.Temperature": {
	  "ambientTemperatureCelsius": 19.73999
	}
  },
  "parentRelations": [
	{
	  "parent": "enterprises/XXXXXXXX/structures/XXXXXXXX/rooms/XXXXXXXX",
	  "displayName": "Upstairs"
	}
  ]
}
```

allenporter committed:
Update nest climate for devices w/o FAN_MODE

mike9011 commented:
> FYI, the European version can do Domestic Hot Water (DHW) but I don't have that enabled for my thermostats at the moment, so there could be other traits that are missing from my output.

I have DHW enabled, but i got the same output. 

However i have some issues on the front-end now. It looks like `target_temp_high: null` and  `target_temp_low: 21.7` are added to the card.
![image](https://user-images.githubusercontent.com/18025163/97963368-480d8b80-1db7-11eb-8897-48307ebdfa6f.png)
![image](https://user-images.githubusercontent.com/18025163/97963591-b3eff400-1db7-11eb-9740-1a7da7b5d61d.png)

allenporter committed:
Update nest climage target temperature logic

This adjuts the temperature logic so that it correctly implements low and high
based on the supported features and hvac mode.

danstreeter commented:
Hi @mike9011 - Sorry to jump in on here, as I've no doubt that its somewhere on the internet... but could you point me to or advise on how I can run this dev version on my HASS instance to:
a) Assist in any additional debugging on this
b) Start getting involved with how HASS development works
c) _Control my nest from HASS!_

balloob commented on homeassistant/components/nest/climate_sdm.py:
```diff
@@ -0,0 +1,331 @@
+    subscriber = hass.data[DOMAIN][entry.entry_id]
+    device_manager = await subscriber.async_get_device_manager()
+
+    # Fetch initial data so we have data when entities subscribe.
```
Is this a to do item?

allenporter replied:
Really this is happening under the covers with the device manager above, so I just removed this comment to reduce confusion.  I think this was copied over from when I built this on poll mode weeks ago.

balloob commented on homeassistant/components/nest/climate_sdm.py:
```diff
@@ -0,0 +1,331 @@
+    def supported_features(self):
+        """Bitmap of supported features."""
+        features = 0
+        if ThermostatTemperatureSetpointTrait.NAME in self._device.traits:
```
Is this something we could calculate inside `async_added_to_hass` ?

allenporter replied:
I am not totally sure I am implementing this correctly, and would like some guidance here.  In particular, the supported mode for target temperature vs target temperature range is based on the current HVAC mode, which is something that will change on update.

I *think* either I am implementing this wrong, and it should support both modes, or it needs to be refreshed any time there is a trait update to check the new mode.   I am not sure i trust the other implementations.

Is it better to always support a range and have a None high/low temp? or only support that feature when in the right mode?   I'd like to write a ClimateEntity spec / documentation after this.

allenporter replied:
I looked at the zha climate implementation and modeled behavior based on that.

balloob replied:
Yeah sorry, it's a mess. This could use some improvements.

balloob reviewed (approved):
Nice! Two tiny comments.

allenporter committed:
Memoize supported_features on device creation

allenporter committed:
Update nest climate target temperature

Update nest climate to implement the climate entity matching how
other integrations do it: the supported modes are always present, then
only populated when appropriate.

balloob commented on homeassistant/components/nest/climate_legacy.py:
```diff
@@ -0,0 +1,355 @@
+
+        if self.device.can_heat and self.device.can_cool:
+            self._operation_list.append(HVAC_MODE_AUTO)
+            self._support_flags = self._support_flags | SUPPORT_TARGET_TEMPERATURE_RANGE
```
```suggestion
            self._support_flags |= SUPPORT_TARGET_TEMPERATURE_RANGE
```

balloob commented on homeassistant/components/nest/climate_legacy.py:
```diff
@@ -0,0 +1,355 @@
+        # feature of device
+        self._has_fan = self.device.has_fan
+        if self._has_fan:
+            self._support_flags = self._support_flags | SUPPORT_FAN_MODE
```
```suggestion
            self._support_flags |= SUPPORT_FAN_MODE
```

balloob reviewed (approved):
🎉 🎉 🎉

allenporter committed:
Simplify supported features

allenporter committed:
SImplify feature flag syntax

allenporter commented:
I applied the last suggestion in a few more places.  I think this is ready to go.

allistermaguire commented:
Hi @allenporter,

I am not sure if I should comment here or raise an issue, but I using this version and it adds an extra mode "auto", when my thermostat only supports "heat" and "off". THERMOSTAT_ECO_MODE mode adds it because it assumes SUPPORT_TARGET_TEMPERATURE_RANGE, and the trait returns both heatCelsius and coolCelsius even thought only heatCelsius is configurable for my thermostat.

Would it make more sense for THERMOSTAT_ECO_MODE mode to map temperatures based on supported modes? If it didn't support HVAC_MODE_HEAT_COOL or HVAC_MODE_HEAT and HVAC_MODE_COOL it would just use a set point instead of a range.

frenck commented:
@allistermaguire Please raise issues instead of commenting on handled PRs. Thanks 👍

MartinHjelmare commented on homeassistant/components/nest/climate_sdm.py:
```diff
@@ -0,0 +1,345 @@
+    def fan_modes(self):
+        """Return the list of available fan modes."""
+        if FanTrait.NAME in self._device.traits:
+            return list(FAN_INV_MODE_MAP.keys())
```
Copying a dict to a list copies the keys by default. We can remove `.keys()`.

MartinHjelmare commented on homeassistant/components/nest/climate_sdm.py:
```diff
@@ -0,0 +1,345 @@
+        trait = self._device.traits[ThermostatTemperatureSetpointTrait.NAME]
+        if self.preset_mode == PRESET_ECO or self.hvac_mode == HVAC_MODE_HEAT_COOL:
+            if low_temp and high_temp:
+                return await trait.set_range(low_temp, high_temp)
```
Nothing is checking this return value. Please separate the await from the return and have the return statement on its own on a separate line.

MartinHjelmare reviewed (commented):
Nice! Just some small clean up for the future.
